front/pages/api/login.ts

```diff
@@ -66,275 +98,281 @@ async function handler(
-          await _createAndLogMembership({
-            workspace,
-            userId: user.id,
-            role: "admin",
-          });
-
-          await internalSubscribeWorkspaceToFreeTestPlan({
-            workspaceId: workspace.sId,
-          });
-          isAdminOnboarding = true;
-        }
-      }
+  // `membershipInvite` is set to a `MembeshipInvitation` if the query includes an
+  // `inviteToken`, meaning the user is going through the invite by email flow.
+  let membershipInvite: MembershipInvitation | null = null;
+
+  if (inviteToken && typeof inviteToken === "string") {
+    const decodedToken = verify(inviteToken, DUST_INVITE_TOKEN_SECRET) as {
+      membershipInvitationId: number;
+    };
+
+    membershipInvite = await MembershipInvitation.findOne({
+      where: {
+        id: decodedToken.membershipInvitationId,
+      },
+    });
+    if (!membershipInvite) {
+      return apiError(req, res, {
+        status_code: 400,
+        api_error: {
+          type: "invalid_request_error",
+          message:
+            "The invite token is invalid, please ask your admin to resend an invitation.",
+        },
+      });
+    }
+    if (membershipInvite.status !== "pending") {
+      membershipInvite = null;
+    }
+  }
 
-      let targetWorkspace: Workspace | null = null;
+  // Login flow: first step is to attempt to find the user.
+  let user = await User.findOne({
+    where: {
+      provider: session.provider.provider,
+      providerId: session.provider.id.toString(),
+    },
+  });
 
-      // `workspaceInvite` flow: we know we can add the user to the workspace as all the checks
-      // have been run. Simply create the membership if does not alreayd exist.
-      if (workspaceInvite) {
-        let m = await Membership.findOne({
-          where: {
-            userId: user.id,
-            workspaceId: workspaceInvite.id,
-          },
-        });
+  // The user already exists, we create memberships as needed given the values of
+  // `workspaceInvite` and `membershipInvite`.
+  if (user) {
+    // Update the user object from the updated session information.
+    user.username = session.user.username;
+    user.name = session.user.name;
 
-        if (!m) {
-          m = await _createAndLogMembership({
-            workspace: workspaceInvite,
-            userId: user.id,
-            role: "user",
-          });
-        }
-
-        if (m.role === "revoked") {
-          return apiError(req, res, {
-            status_code: 400,
-            api_error: {
-              type: "invalid_request_error",
-              message:
-                "Your access to the workspace has been revoked, please contact the workspace admin to update your role.",
-            },
-          });
-        }
-
-        targetWorkspace = workspaceInvite;
-      }
+    // We only update the user's email if the session is not from Google.
+    if (!isGoogleSession(session)) {
```

Comment:
why? is it possible for a google user to change their email? if yes shouldn't we update?

Reply:
Based on what I've gathered from their documentation, it seems that it isn't feasible. This being said, I don't see why we should update user's email once they have been created. The only exception is for GitHub since they let users change their email address.